boring/src/aead.rs

```diff
@@ -0,0 +1,357 @@
+//! High level interface to certain AEAD schemes.
+//!
+//! # Examples
+//!
+//! Encrypt data with AES128-GCM-SIV.
+//!
+//! ```
+//! use boring::aead::{encrypt_aead, Aead};
+//!
+//! let aead = Aead::aes_128_gcm_siv();
+//! let data = b"Some Crypto Text";
+//! let aad = b"Some Crypto Context";
+//! let key = b"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F";
+//! let nonce = b"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x10\x11";
+//! let sealed_data = encrypt_aead(
+//!     aead,
+//!     key,
+//!     nonce,
+//!     aad,
+//!     data).unwrap();
+//!
+//! assert_eq!(
+//!     b"\x3D\xCC\xF8\x9A\x2B\xC1\x91\x15\xCD\x19\xD2\xC2\xB1\x13\x01\x38\xC5\
+//!       \x7D\xF5\xEE\x5B\x37\xD1\x44\x05\xD2\x1A\xDA\x3F\x69\x97\x5D",
+//!     &sealed_data[..]);
+//! ```
+
+use crate::ffi;
+use libc::c_int;
+
+use crate::error::ErrorStack;
+use crate::{cvt, cvt_p};
+
+use ffi::{
+    EVP_AEAD_CTX_new, EVP_AEAD_key_length, EVP_AEAD_max_overhead, EVP_AEAD_max_tag_len,
+    EVP_AEAD_nonce_length,
+};
+
+/// Represents a particular AEAD algorithm.
+///
+/// For more information see the BoringSSL documentation on [`aead.h`].
+///
+/// [`aead.h`]: https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Aead(*const ffi::EVP_AEAD);
+
+impl Aead {
+    pub fn aes_128_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_gcm()) }
+    }
+
+    pub fn aes_192_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_192_gcm()) }
+    }
+
+    pub fn aes_256_gcm() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_gcm()) }
+    }
+
+    pub fn chacha20_poly1305() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_chacha20_poly1305()) }
+    }
+
+    pub fn xchacha20_poly1305() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_xchacha20_poly1305()) }
+    }
+
+    pub fn aes_128_ctr_hmac_sha256() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_ctr_hmac_sha256()) }
+    }
+
+    pub fn aes_256_ctr_hmac_sha256() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_ctr_hmac_sha256()) }
+    }
+
+    pub fn aes_128_gcm_siv() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_128_gcm_siv()) }
+    }
+
+    pub fn aes_256_gcm_siv() -> Aead {
+        unsafe { Aead(ffi::EVP_aead_aes_256_gcm_siv()) }
+    }
+
+    /// Returns the length of keys used with this AEAD.
+    pub fn key_len(&self) -> usize {
+        unsafe { EVP_AEAD_key_length(self.0) as usize }
+    }
+
+    /// Returns the length of nonces used with this AEAD.
+    pub fn nonce_len(&self) -> usize {
+        unsafe { EVP_AEAD_nonce_length(self.0) as usize }
+    }
+
+    /// Returns the maximum ciphertext overhead with this AEAD.
```

Comment:
Let's use the upstream documentation here, I'd prefer not to reuse the word "overhead" when defining what it means.
```suggestion
    /// Returns the maximum number of additional bytes added by the act of sealing data with `self`.
    ///
    /// Corresponds to [`EVP_AEAD_max_overhead`](https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html#EVP_AEAD_max_overhead).
```